Fix/propagate optimisation conditions

junniest opened the pull request:
Propagating optimisation conditions from initial settings to optimiser setup (e.g. epsilon)

junniest committed:
Add StopCondition predicate to optimisers/mod

junniest committed:
Replace TopologyOptimiserPredicate w StopCondition

Replaced TopologyOptimiserPredicate use in TopologyOptimiser with StopCondition,
working on unifying stop condition handling for optimisation in all modules.

junniest committed:
Fix topoOptimiser tests for StopCondition

Fixed TopologyOptimiser tests to use the StopCondition predicate, added
a test for max iterations.

junniest committed:
Update BranchOptimiser to use StopCondition

junniest committed:
Fix test names to remove redundancy

junniest committed:
Individual blen optimisations capped at 10 iters

Individual branch length optimisations capped at 10 iters since we are doing multiple
iterations over all branches anyway. StopCondition predicate does not apply to argmin
value optimisation.

junniest committed:
Add StopCondition tests to BranchOptimiser

junniest committed:
Use DEFAULT_EPSILON const for initialisation

junniest committed:
Add `max_iter()` to `StopCondition`

junniest committed:
Remove commented out old impl

junniest committed:
Remove redundant `StopCondition::met()`

junniest committed:
Switch bench to StopCondition

junniest committed:
Rename `new_with_pred` to `with_stop_condition`

Rename `TopologyOptimiser::new_with_pred` to `with_stop_condition`

junniest committed:
Fix TopoOpt precision in tests

junniest committed:
Add missing field to PhyloOptResult in test macro

junniest committed:
Remove invalid check from test

junniest committed:
Display impl for StopCondition

junniest committed:
Add Clone and Debug to `OptimisationResult`s

junniest committed:
Add StopCondition output to optimisers

Add StopCondition output to branch length and topology optimisers to make sure its
clear what the optimisation end condition is

junniest committed:
Remove unnecessary parameter in optimise_iteration

Remove unnecessary parameter from TopologyOptimiser::optimise_iteration, it used to take
the current cost, now it's computed/read from the cost function

junniest committed:
Fix and clarify pip_optimise_model_tree test

The tree w/o model optimisation and the tree with model optimisation used to match exactly, now
the rf distance is 2 for unclear reasons but the likelihoods are improving which is the most
important part. Added comments to the test and renamed vars to make the pipeline clearer.

junniest committed:
Refactor ModelOptimiser to match others

Refactor ModelOptimiser to match the other optimisers: change run() to be shorter and just
set up iterations, optimisation logic split into optimise_frequencies and single_optimisation_iteration.
Added logging of stop condition.

junniest committed:
Rename optimise_iter to single_optimisation_iter

Rename optimise_iteration to single_optimisation_iteration for clarity

junniest committed:
Add costs vec to ModelOptimisationResult

Add costs vec to ModelOptimisationResult to keep track of all costs during the course
of optimisation, mainly for debugging purposes.

junniest committed:
Rename PhyloOptRes to TreeOptimisationResult

Rename PhyloOptimisationResult to TreeOptimisationResult for clarity

junniest committed:
Fix docstring on TopologyOptimiser::run()

junniest committed:
Fix topo optimiser log message

junniest committed:
Fix inconsistent saving of costs in ModelOptimiser

Fix inconsistent saving of costs in ModelOptimiser -- after adding a costs: Vec<f64> field
to ModelOptimisationResult, the first cost pushed into the vector was not the cost at the start
but rather the cost after the potential frequency optimisation which was not consistent behaviour
with other optimisers. Also added the check for the first cost value to tests.

junniest committed:
Fix use statement for clippy

junniest committed:
Fix use statement order

junniest committed:
Use FakeRng to build nj trees for stop cond tests

Using FakeRng in tests to deterministically build nj trees for stopping condition tests
in TopologyOptimiser

junniest committed:
Adding brent optimisation iteration control

Adding brent optimisation max iteration count as a separate configuration setting for branch length
optimiser. The Stop condition applies to the whole optimiser interation (over all branches) whereas it is
helpful to limit the individual branch optimisation to e.g. 10 iterations of Brent max, since this is used
to check each individual possible move and otherwise takes too long.

junniest committed:
Add a way to set max_brent_iters

junniest committed:
Add Default to StopCondition and use everywhere

junniest committed:
Fix comments in TopologyOptimiser tests

junniest committed:
Move single branch optim out of BranchOptimiser

Moved single branch optimisation out of BranchOptimiser so it can be called by nni and spr
moves without having to set up the BranchOptimiser struct.

junniest committed:
Add debug_assert to check costs match

Add debug_assert to end of optimisation cycles in Branch- and TopoOptimisers to check that
the cost from the single iteration matches the cost stored within the optimiser.
Remove duplicate assertion after the cycle is finished.

junniest committed:
Add single_iter tests to BranchOptimiser

junniest committed:
Add single_iter tests to TopoOptimiser

junniest committed:
Add a wrong test tree for protein example

junniest committed:
Remove max brent iterations from spr optimiser

Remove max brent iterations from spr optimiser because it was causing the debug_assert!
in topo_optimiser::run() to fail. Still no idea why that is the case.

junniest committed:
Fix stop condition tests to use methods

Fixed stop condition tests to use methods instead of instantiating the enum directly

junniest committed:
Fix stop cond tests to remove not from check

Fixed stop condition tests to remove not from check, used to accept the opposite of the expected
value.

junniest committed:
Add test for StopCondition::max_iter

junniest committed:
Add test for StopCondition::display()

junniest committed:
Remove wrong ignore flag from fast tests

junniest committed:
Add test for Estimated freqs default to empirical

Added test for FrequencyOptimisation::Estimated freqs defaulting to empirical,
will become obsolete when estimation is implemented.

junniest committed:
Add test for ModelOptimiser::with_stop_condition()

junniest committed:
Remove StopCondition override in topo_opt tests

Removed unnecessary StopCondition override in topo_opt tests

junniest committed:
Fix param order in ModelOptimiser::with_stop_cond

Fix parameter order in ModelOptimiser::with_stop_condition to match other optimisers.
StopCondition should come last.

MattesMrzik commented on phylo/src/optimisers/model_optimiser.rs:
```diff
@@ -7,86 +7,116 @@ use log::{debug, info, warn};
+        cost: C,
+        stop_condition: StopCondition,
+        freq_opt: FrequencyOptimisation,
+    ) -> Self {
+        Self {
+            stop_condition,
             c: cost,
             freq_opt,
         }
     }
 
     pub fn run(mut self) -> Result<ModelOptimisationResult<C>> {
         info!("Optimising the evolutionary model: {}", self.c);
+        info!("Optimisation stopping condition: {}", self.stop_condition);
 
         let init_cost = self.c.cost();
         info!("Initial cost: {init_cost}");
-        let mut curr_cost = init_cost;
+
+        let mut curr_cost = self.optimise_frequencies();
+        debug_assert!(curr_cost >= init_cost);
```
This comment is wrong since we do 
```rust
let init_cost = self.c.cost();
curr_cost = self.optimise_frequencies();
let mut costs = vec![init_cost, curr_cost];
```

MattesMrzik commented on phylo/src/optimisers/model_optimiser.rs:
```diff
@@ -134,3 +165,84 @@ impl<C: ModelSearchCost> CostFunction for ParamOptimiser<C> {
+
+        let model = SubstModel::<GTR>::new(&[], &[]);
+        let cost = SCB::new(model, info).build().unwrap();
+        let mut opt = ModelOptimiser::new(cost.clone(), FrequencyOptimisation::Estimated);
```
Why not use `FrequencyOptimisation::Empirical` here? The tests are called ...empirical... and you assert against `&opt.c.empirical_freqs()`.
(See also the tests below)

junniest replied:
These are actually checking that the `FrequencyOptimisation::Estimated` variant ends up defaulting to empirical frequencies, since proper ML estimation is not actually implemented. These will get removed once the estimation is implemented properly, it was just to make sure that the behaviour is meaningful! The name of this test is actually `estimated_freqs_default_to_empirical_dna`, apologies for the confusion!

MattesMrzik commented on phylo/src/optimisers/topo_optimiser_tests.rs:
```diff
@@ -481,55 +481,60 @@ fn pip_optimise_model_tree() {
+    assert!(model_tree_opt.final_cost > tree_opt.final_cost);
+
+    // Check that the optimised tree+model is better than the starting tree with the optimised model
+    let old_tree_optimised_model_logl = PIPCB::new(optimised_pip, tree_opt.cost.info.clone())
```
How is this the old tree if you use `tree_opt` which already did optimize the tree? Is it because you run the topo opti twice and you are checking the final tree vs the second tree?

junniest replied:
This is indeed confusing... What I tried to say is that the `old_tree` is the tree optimised without model optimisation and if we use the optimised model to compute the likelihood it should still be better. I will rename this to `tree_opt` to match the naming scheme in the rest of the test. I don't think this is a good naming scheme anyway, but that was the best I could come up with to try and distinguish between different objects.

MattesMrzik reviewed (approved):
Good PR

junniest committed:
Rename/fix comments on one of the confusing tests

Renamed intermediate variables and clarified the comments in one of the more confusing optimisation
tests.

junniest committed:
Mark some longer tests ignored for coverage runs
